src/compiler/scala/tools/nsc/PhaseAssembly.scala

```diff
@@ -224,38 +234,92 @@ object DependencyGraph {
+  def apply(components: Iterable[SubComponent], warnAll: Boolean = true): DependencyGraph = {
+    val phases = components.toArray
+    val phaseNames = components.iterator.map(_.phaseName).toSet
+    def isPhaseName(s: String): Boolean = phaseNames.contains(s)
     val start = phases.find(_.initial)
       .orElse(phases.find(_.phaseName == Parser))
       .getOrElse(throw new AssertionError("Missing initial component"))
     val end = phases.find(_.terminal)
       .orElse(phases.find(_.phaseName == Terminal))
       .getOrElse(throw new AssertionError("Missing terminal component"))
-    new DependencyGraph(phases.size, start.phaseName, phases.map(p => p.phaseName -> p).toMap).tap { graph =>
-      for (p <- phases) {
-        require(p.phaseName.nonEmpty, "Phase name must be non-empty.")
-        def checkConstraint(name: String, constraint: String): Boolean =
-          graph.components.contains(name).tap(ok => if (!ok) graph.warning(s"No phase `$name` for ${p.phaseName}.$constraint"))
-        for (after <- p.runsRightAfter if after.nonEmpty && checkConstraint(after, "runsRightAfter"))
-          graph.addEdge(after, p.phaseName, FollowsNow)
-        for (after <- p.runsAfter if after.nonEmpty && !p.runsRightAfter.contains(after) && checkConstraint(after, "runsAfter"))
-          graph.addEdge(after, p.phaseName, Follows)
-        for (before <- p.runsBefore if before.nonEmpty && checkConstraint(before, "runsBefore"))
-          graph.addEdge(p.phaseName, before, Follows)
-        if (p != start && p != end)
-          if (p.runsRightAfter.forall(_.isEmpty) && p.runsAfter.forall(_.isEmpty))
-            graph.addEdge(start.phaseName, p.phaseName, Follows)
-        if (p != end || p == end && p == start)
-          if (p.runsBefore.forall(_.isEmpty))
-            graph.addEdge(p.phaseName, end.phaseName, Follows)
+    val constraints = ListBuffer.empty[(String, String, Weight)]
+    def addConstraint(from: String, to: String, weight: Weight): Unit = constraints.addOne((from, to, weight))
+    val warnings = ListBuffer.empty[String]
+    var i = 0
+    while (i < phases.length) {
+      val p = phases(i)
+      require(p.phaseName.nonEmpty, "Phase name must be non-empty.")
+      var constrained = false
+      var badConstraint = false
+      def checkConstraint(name: String, constraint: String, warn: Boolean): Boolean = isPhaseName(name).tap { ok =>
+        if (!ok && warn) {
+          val help = if (compilerPhases.contains(name)) "" else
+            compilerPhases.filter(util.EditDistance.levenshtein(name, _) < 3) match {
+              case Nil => ""
+              case close => s" - did you mean ${util.StringUtil.oxford(close, "or")}?"
+            }
+          warnings.addOne(s"No phase `$name` for ${p.phaseName}.$constraint$help")
+          badConstraint = true
+        }
       }
+      // add edges for each constraint
+      for (after <- p.runsRightAfter if after.nonEmpty && checkConstraint(after, "runsRightAfter", warn=true)) {
```

Comment:
I found one more corner case... 2.13.14 initially adds the `platformPhases` when running Scaladoc, then they are silently dropped when the algorithm runs. But that affects the behavior for a plugin phase with `runsAfter flatten`: it's dropped without warning.

There's also Sebastien's report that 2.13.14 Scaladoc drops Scala.js phases without warning (https://github.com/scala/scala/pull/10856#discussion_r1744224563).

I assume the warning in 2.13.14 for dropping a phase is buggy / not always issued when it "should" have been.

So should we do `warn = warnAll` here?

Reply:
I remember that the weird construction of phases was why I wanted to "nail it down".

For the warning, I think we can design the behavior. My basis for not warning is that people have arbitrary plugins enabled for scalac in their build, and scaladoc (as a doc task) just inherits that cruft.

This makes me think that scaladoc should just be a plugin.

I haven't done another pass on the PR yet, but oh I see you're saying don't warn here under scaladoc. I bet you're right.